### PR TITLE
feat: Speed predictor datavalue fetching

### PR DESCRIPTION
**Pull request by jimgrace**
This speeds up the fetching of input data from the data value table. In tests with the South Africa malaria database it makes it about 145 times faster. On a test system with PostgreSQL 13, a predictor over three orgUnit levels took 13.8 minutes before the PR, and 5.7 seconds after, returning the same data. This speed improvement is for two reasons:

1. A single SQL query is used to fetch all datavalue table input data, rather than one query per 500 orgUnits (resulting in 54 queries for the test case). Data is pipelined row-by-row from the query results to the prediction logic using a callback routine, with the result that datavalues for only one orgUnit needs to be in memory at any one time, instead of datavalues for 500 orgUnits.

Analytics data fetching is the same as before--data is fetched for 500 orgUnits at a time. To handle this logic, separate classes were made for providing aggregate datavalues vs. analytics data, so that each can be optimized differently. A third new class was added for writing predictions (adding, updating, (soft) deleting any existing predictions as needed, or leaving them  unchanged). This further simplifies the logic in the main prediction service:
```
  DefaultPredictionService
    |- PredictionDataValueFetcher (pipelined data through a callback)
    |- PredictionAnalyticsDataFetcher (500 orgUnits at a time)
    |- PredictionWriter (merges with existing predicted values)

```
2. In the SQL query, instead of joining the organisationunit table twice to the datavalue table, it is only joined once. It had been joined twice to fetch data for all orgUnit descendants: one instance was used to select the orgUnits for which data was to be returned (trimmed to only those orgunits the user is allowed to see from their assigned orgunits.) A second instance was joined with the first one to find all descendants of those orgUnits, and joined to the datavalue table to find data for those descendants.

The query was modified to join the organisationunit table only once. OrgUnit selection is done through organisationunit.hierarchylevel, since we are now selecting all the orgUnits at a level and not just 500 at a time. OrgUnit selection is further restricted to only what the user can see by checking that the orgUnit path starts with the path of the orgUnit(s) assigned to the user. Values from descendant orgUnits are selected by requiring the same path restrictions, and looking for the hierarchylevel greater than or equal to the data sampling level.

### SonarCloud Note:
The PredictionDataValueFetcher uses a second thread to fetch datavalues. If an unexpected RuntimeException were to occur on that thread and it wasn't caught, the primary thread would wait on the semaphore for data to be ready and eventually time out and report the timeout exception. I thought it would be better for the primary thread to report the real exception.

So what the code does is catch any unexpected RuntimeException on the second thread and store that exception in a variable, set the semaphore that data is ready, and exit the thread. When the primary thread finds data ready, it first looks at the exception variable and re-throws it if it's there (not null). The re-thrown exception will contain a combined stack trace of the primary thread, and below it the stack trace within the secondary thread. That seemed to me the best way to show what happened, and tie the exception to the call on the primary thread under which it occurred. I've tested it and it works that way. (There's even a unit test that throws a NPE on the second thread by way of a mock and checks that it is re-thrown by the primary thread.)

SonarCloud doesn't like that the exception is not re-thrown in the second thread after it is caught. It's been marked as "Won't Fix".

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/datavalue/hibernate/HibernateDataValueStore.java
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/datavalue/hibernate/HibernateDataValueStore.java:331]: Could we move this check to the service layer potentially? Ideally validation should be in one place. We can also use an `ErrorCode` for programmatic handling.

Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/datavalue/hibernate/HibernateDataValueStore.java]: Thanks. Done!

**Comment by stian-sandvold** [comment]
@jimgrace there is a potential bug reported from sonarqube, could you take a look at it?
https://sonarcloud.io/project/issues?id=dhis2_dhis2-core&issues=AXdViYO0b1S7AK3UFFna&open=AXdViYO0b1S7AK3UFFna&pullRequest=7214

**Comment by jimgrace** [comment]
Hi @stian-sandvold, thanks. I acknowledged that SonarCloud report in the PR description, but my code works as intended.

To be more specific: I'm using a second thread to fetch datavalues. If an unexpected RuntimeException were to occur on that thread and it wasn't caught, the primary thread would wait on the semaphore for data to be ready and eventually time out and report the timeout exception. I thought it would be better for the primary thread to report the real exception.

So what I do is catch any unexpected RuntimeException on the second thread and store that exception in a variable, set the semaphore that data is ready, and exit the thread. When the primary thread finds data ready, it first looks at the exception variable and re-throws it if it's there (not null). The re-thrown exception will contain a combined stack trace of the primary thread, and below it the stack trace within the secondary thread. That seemed to me the best way to show what happened, and tie the exception to the call on the primary thread under which it occurred. I've tested it and it works that way. (There's even a unit test that throws a NPE on the second thread by way of a mock and checks that it is re-thrown by the primary thread.)

If we want to keep SonarCloud happy, I could save the exception in a variable to be re-thrown, set the semaphore that data is ready and also re-throw the exception there on the second thread. That way the exception would get logged the way it does now showing the stack from both threads, and it would also get logged a second time under the second thread with only the second thread stack. IMHO it's not as nice, but it would keep SonarCloud happy. That's where I'm open to suggestions as to what to do.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java
Comment (aimiebell) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java:622]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [RC: Suspicious reference comparison of Boolean values (RC_REF_COMPARISON_BAD_PRACTICE_BOOLEAN)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rc-suspicious-reference-comparison-of-boolean-values-rc-ref-comparison-bad-practice-boolean).
This method compares two Boolean values using the == or != operator. Normally, there are only two Boolean values (Boolean.TRUE and Boolean.FALSE), but it is possible to create other Boolean objects using the new Boolean(b) constructor. It is best to avoid such objects, but if they do exist, then checking Boolean objects for equality using == or != will give results than are different than you would get using .equals(...).

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataExportParams.java
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataExportParams.java:406]: I would prefer not to have a specific "orgUnitParents" collection, and instead just have the existing org unit collection and in addition include an "org unit selection mode" property.

We have an `OrganisationUnitSelectionMode` enum in the system, where we have relevant values like `SELECTED` and `DESCENDANTS`. `SELECTED` is default and current behavior. `DESCENDANTS` means include the sub-hierarchy. This is more flexible as it allows for more modes/options.

See event analytics as an example, scroll to "ouMode":
 
https://docs.dhis2.org/master/en/dhis2_developer_manual/web-api.html#event-analytics

Sounds reasonable?

Comment (jimgrace) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DataExportParams.java]: Makes sense. Done!

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/datavalue/hibernate/HibernateDataValueStore.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/datavalue/hibernate/HibernateDataValueStore.java]: Overcome by events. In the later refactor to use OrganisationUnitSelectionMode, this error check was removed.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DeflatedDataValueHandler.java
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DeflatedDataValueHandler.java:35]: These usually are called `Consumer` in the JRE, maybe we should continue that if they are general purpose and do not have any "event" handler characteristic which springs to mind when using the suffix `Handler`.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java:520]: A way to avoid multiple `continue`s would be to extract this to a helper method. I'd guess we do something similar elsewhere too.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java:66]: `private`?

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java:223]: Usually java projects try to not use `continue`. I don't know if we follow this trend. In this case for example I'd rather move the 3 lines below into the `if` and negate it.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:316]: There is quite some duplication here making this unnecessarily long and hard to see if or what the varying value(s) are for the reoccurring calls.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:340]: I feel like each size should be its own test instead of switch

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionDataValueFetcherTest.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionDataValueFetcherTest.java:81]: There are several pattern of duplication in this class, could do with some helper methods.

**Review by jbee — COMMENTED** [review]
Good work with commenting crucial methods with reasons and reasoning. 

I have not been in that area of the project before. Overall I got the impression the "low level DB" is leaking into layers above a bit too much blurring the lines between working with a data model and working with database level concerns. 

When I was reading your description on the thread synchronisation I was thinking this sounds like the perfect job to use a `CompletableFuture` for synchronisation. This captures the `Future<List<T>>` in case of success but you could also complete with an exception which is then re-thrown (wrapped) for the called of the `Future.get` (also allows timeout). You might consider using that instead of a semaphore and a hand crafted passing and checking of exceptions.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DeflatedDataValueHandler.java
Comment (jimgrace) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/datavalue/DeflatedDataValueHandler.java]: I had been thinking of "handle" in the most generic sense I could imagine -- a class that can accept a _DefaltedDataValue_ and do something with it, "handle" it. But I agree that handle may be too closely associated with events, so I've changed it as you suggest to consumer.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java]: Great idea. Done.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java]: Yup, changed. Thanks.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java]: If I look for `continue;` in the project, I see a lot of them. But I've removed this one as you suggested. I try not to let my blocks get too long if I can do an early return/continue, but 3 statements is not too long! Good suggestion.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:324]: I'm of two minds about this. I agree that helper methods are useful when it makes the intent of the code clearer and I don't mind using them when I think they do that. But when a test is failing, someone new to the test code needs to know exactly what the test conditions are. Sometimes code that is a bit less optimized may be easier to see exactly what is being instrumented. I find in general that a lot of DHIS2 unit tests have a lot of copy-and-paste code that would be totally unacceptable in operational code.

I don't think that helper routines should never be used to optimize copy-and-paste style test code, nor that they should always be used. It's a value judgement each time, where the end goal is always how quickly can someone not familiar with the test code know exactly what is going on.

In this particular case, we're building a lot of _DataQueryParams_ objects, which have their own builder, and _Grid_ objects to which we add rows. I considered your suggestion, and I tried to imagine how I would design them. It's not the case that most of the parameter never change. They change at different times. In this particular case, I felt that helper methods might make it less clear to the reader exactly how these objects are being constructed. We don't want the reader spending a lot of time wondering if an error is in the operational code or in a test helper.

I would also be very interested in any opinions from other group members, or in any further thoughts from you.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:352]: I sympathize. What I wanted to do here is to fetch every one of the 1005 orgUnits in order, because that is how the PredictionAnalyticsDataFetcher is designed to be used, and be sure that the results were where they were should be, and not elsewhere. This seemed to me to be a reasonable way to do that.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionDataValueFetcherTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionDataValueFetcherTest.java:81]: Understood. See the comment above.

**Comment by jimgrace** [comment]
@jbee Thank you for taking the time and making some good comments on the code!

Regarding data model vs. DB: I believe that whenever possible we work with the data model. But when performance is critical, we use SQL. So yes, in those cases the DB does leak up a bit. Does this make sense? Or perhaps I'm missing an opportunity to understand better what you are saying.

Regarding using _CompletableFuture_, that might be a really great idea. I've worked a bit with _Promises_ in JavaScript but haven't yet used _Futures_ in Java. For the task at hand, one of my objectives was to avoid a disruption to the way HibernateDataValueStore.getDeflatedDataValues works, since it is also used by validation rules. It seemed that a "callback" object for predictors would be an easy option to implement, while allowing the data to be collected and returned in a more conventional way for validation rules. But for the next release we might consider using the pipelined data approach for validation rules as well, and then perhaps we could look at a common approach that uses _CompletableFuture_ for both.

On a practical note, this code has been available for review for 24 days. We have 2 days left until the soft feature freeze for 2.36, and I am hoping to get this code into that release.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jbee) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:324]: I understand what you say and I am not here to argue with you or ask for changes that you don't see as an improvement.

I know there is similar duplication in many tests but I in my experience this is not a good thing. I have cleaned up several of these and found that when there is repetitive code there are hidden issues because it gets hard to see the forest for the trees. 

Just now I work on straightening out transaction management and what I especially find with tests repeating basic steps instead of using helpers is that tests do things in the wrong order (use not yet saved entities). This then just works because of hibernate magic queuing the missed operations in the background and squeezing them in where you would not expect them. Except that you might be in a read-only transaction by then. It appears in some places the solution to this has not been to straighten out the test setup but to make read-only DB access use a read-write transaction. This then cascades up and soon lots of things have to be read-write although they clearly only want to read from the database. 

I also can't say that lengthy code helps me understanding what is going on. From a certain length beyond 20-25 lines   instead of being able to read the declared intent of what the test scenario is trying to do there are all these details making me wonder what the tests tries to do, what the important and unimportant values are and so on. It basically means the only way to understand the test is to reverse engineer it and assume intent instead of reading intent and then in case something is odd see if the actual code fails to fulfil what was intended to happen. 
I miss the WHY and the WHAT. There is lots of HOW for some unknown WHAT and WHY. This is what I think helper methods can provide when named accordingly.

**Review by jbee — APPROVED** [review]
@jimgrace Thanks for making the changes. Since my comments where purely about code style I feel someone that knows the semantics of the code better than me should have a look before you merge it.

**Comment by jimgrace** [comment]
@jbee Thanks. I appreciate the review and the good comments.

I've pretty much written and maintained all of the current _DefaultPredictionService_ code and also the _getDeflatedDataValues_ method in _HibernateDataValueStore_ that it calls (and to which I've added the callback). The people besides myself who I'm guessing have seen the code the most are Lars, who has commented on but not approved this PR, and Stian, who has approved it.

I'd be very happy to get more reviews and approvals, but I don't know if I will get them. I may go with the approvals I have by tomorrow's code freeze.

@larshelge Any further thoughts on this PR?

**Comment by jbee** [comment]
@jimgrace not sure why CodeQL fails - maybe @larshelge can merge anyway.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java:324]: Thanks. Points well-taken about our tests. I've had the same problem. Perhaps we can do better in the future.

I also think that our code in general is under-commented. I know this is an art where good programmers can have a difference of opinion; on the one hand not wanting comments to get in the way of reading code -- but on the other hand wanting to explain the non-obvious things that some code blocks are doing. I tend to comment more rather than less, but I have cut back in this project because not everyone feels the same. I've tended to put JavaDoc on all private methods I've written as a general rule, because that's less obtrusive to some than interspersing them with the code.

You make a good point about choosing names in the code to make the intent clearer. That can be a good middle ground. I can try to do that more. And I think I may also add more comments especially in my tests in the future to make the WHAT and WHY clearer.

At the very least I'd like to see JavaDoc on all our public classes and methods, rather than having to read all the code to figure out their intent.

**Pull request merged**
